src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp

```diff
@@ -0,0 +1,103 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+\*******************************************************************/
+#include "dfcc_infer_loop_assigns.h"
+
+#include <util/expr.h>
+#include <util/find_symbols.h>
+#include <util/message.h>
+#include <util/pointer_expr.h>
+#include <util/std_code.h>
+
+#include <goto-instrument/havoc_utils.h>
+#include <langapi/language_util.h>
+
+#include "dfcc_root_object.h"
+
+/// Builds a call expression `object_whole(expr)`
+static exprt
+make_object_whole_call_expr(const exprt &expr, const namespacet &ns)
+{
+  const symbolt &object_whole_sym = ns.lookup(CPROVER_PREFIX "object_whole");
+  const code_typet &object_whole_code_type =
+    to_code_type(object_whole_sym.type);
+  exprt::operandst args{{expr}};
+  return side_effect_expr_function_callt(
+    object_whole_sym.symbol_expr(),
+    args,
+    object_whole_code_type.return_type(),
+    expr.source_location());
+}
+
+/// Returns true iff \p expr contains at least one identifier found in
+/// \p identifiers.
+static bool
+depends_on(const exprt &expr, std::unordered_set<irep_idt> identifiers)
+{
+  const std::unordered_set<irep_idt> ids = find_symbol_identifiers(expr);
+  for(const auto &id : ids)
+  {
+    if(identifiers.find(id) != identifiers.end())
+      return true;
+  }
+  return false;
+}
+
+std::set<exprt> dfcc_infer_loop_assigns(
+  const local_may_aliast &local_may_alias,
+  const loopt &loop_instructions,
+  const source_locationt &loop_head_location,
+  messaget &log,
+  const namespacet &ns)
+{
+  // infer
+  assignst assigns;
+  get_assigns(local_may_alias, loop_instructions, assigns);
+
+  // compute locals
+  std::unordered_set<irep_idt> loop_locals;
+  for(const auto &target : loop_instructions)
+  {
+    if(target->is_decl())
+      loop_locals.insert(target->decl_symbol().get_identifier());
+  }
+
+  // widen or drop targets that depend on loop-locals or are non-constant
+  havoc_utils_is_constantt is_constant(assigns, ns);
+  assignst result;
+  for(const auto &expr : assigns)
+  {
+    if(depends_on(expr, loop_locals))
+    {
+      // Target depends on loop locals, attempt widening to the root object
+      auto root_object = dfcc_assigns_target_single_root_object(expr, ns, log);
+      if(!depends_on(root_object, loop_locals))
+      {
+        address_of_exprt address_of_root_object(root_object);
+        address_of_root_object.add_source_location() =
+          root_object.source_location();
+        result.emplace(make_object_whole_call_expr(address_of_root_object, ns));
+      }
+    }
+    else
+    {
+      address_of_exprt address_of_expr(expr);
+      address_of_expr.add_source_location() = expr.source_location();
+      if(!is_constant(address_of_expr))
+      {
+        // Target address is not constant, widening to the whole object
+        result.emplace(make_object_whole_call_expr(address_of_expr, ns));
```

Comment:
What happens behind the scenes with a pointer passed to object-whole? Would it be the same as the `root_object` case above?

Reply:
The meaning of  `object_whole(&expr)` or `object_whole(&root_object(expr))` is the same so we don't bother computing the root object. In the first case we compute the root_object to try and get rid of a dependency on loop locals or drop the target if the root object is still local to the loop. In the second case we have a non-constant expression that does not depend on loop locals, and we just widen it directly.